File: back/src/main/java/com/back/domain/trend/npmi/NpmiCalculator.java
package com.back.domain.trend.npmi;

import com.back.domain.trend.npmi.dto.CooccurrenceStatsDto;
import org.springframework.stereotype.Component;

@Component
public class NpmiCalculator {

    public double calculate(CooccurrenceStatsDto stats) {
        if (stats.totalMessages() <= 0) {
            return 0.0;
        }
        if (stats.freqXY() == 0) {
            return -1.0;
        }
        double pX = calculateProbability(stats.freqX(), stats.totalMessages());
        double pY = calculateProbability(stats.freqY(), stats.totalMessages());
        double pXY = calculateProbability(stats.freqXY(), stats.totalMessages());
        double pmi = calculatePmi(pXY, pX, pY);
        return -pmi / (Math.log(pXY) / Math.log(2));
    }

    private double calculateProbability(long freq, long totalMessages) {
        return (double) freq / totalMessages;
    }

    private double calculatePmi(double pXY, double pX, double pY) {
        return Math.log(pXY / (pX * pY)) / Math.log(2);
    }
}

File: back/src/main/kotlin/com/back/domain/trend/npmi/NpmiCalculator.kt
package com.back.domain.trend.npmi

import com.back.domain.trend.npmi.dto.CooccurrenceStatsDto
import org.springframework.stereotype.Component
import kotlin.math.ln

@Component
class NpmiCalculator {
    companion object {
        private val LOG_2 = ln(2.0)
    }

    fun calculate(stats: CooccurrenceStatsDto): Double {
        if (stats.totalMessages <= 0) {
            return 0.0
        }
        if (stats.freqXY == 0L) {
            return -1.0
        }
        if (stats.freqXY == stats.totalMessages) {
            return 1.0
        }

        val pX = calculateProbability(stats.freqX, stats.totalMessages)
        val pY = calculateProbability(stats.freqY, stats.totalMessages)
        val pXY = calculateProbability(stats.freqXY, stats.totalMessages)
        val pmi = calculatePmi(pXY, pX, pY)
        return -pmi / (ln(pXY) / LOG_2)
    }

    private fun calculateProbability(freq: Long, totalMessages: Long): Double = freq.toDouble() / totalMessages

    private fun calculatePmi(pXY: Double, pX: Double, pY: Double): Double = ln(pXY / (pX * pY)) / LOG_2
}
